File: ios/CareLog/CareLog/Voice/VoiceAcknowledgementPlayer.swift
import AVFoundation
import Foundation

/// Plays voice acknowledgement audio clips for vital logging events.
///
/// Audio plays through media stream to work even in silent mode.
final class VoiceAcknowledgementPlayer {
    static let shared = VoiceAcknowledgementPlayer()

    private var audioPlayer: AVAudioPlayer?

    private init() {
        setupAudioSession()
    }

    // MARK: - Audio Session Setup

    private func setupAudioSession() {
        do {
            // Configure audio session to play through speaker even in silent mode
            try AVAudioSession.sharedInstance().setCategory(
                .playback,
                mode: .default,
                options: [.mixWithOthers]
            )
            try AVAudioSession.sharedInstance().setActive(true)
        } catch {
            print("Failed to configure audio session: \(error)")
        }
    }

    // MARK: - Playback Methods

    /// Play success acknowledgement for a specific vital type.
    func playSuccess(type: ObservationType) {
        let audioName = getSuccessAudioName(for: type)
        playAudio(named: audioName)
    }

    /// Play generic success acknowledgement.
    func playGenericSuccess() {
        playAudio(named: "success_generic")
    }

    /// Play failure acknowledgement.
    func playFailure() {
        playAudio(named: "failure")
    }

    /// Play no network acknowledgement.
    func playNoNetwork() {
        playAudio(named: "no_network")
    }

    /// Play upload success acknowledgement.
    func playUploadSuccess() {
        playAudio(named: "upload_success")
    }

    // MARK: - Private Methods

    private func playAudio(named name: String) {
        guard let url = Bundle.main.url(forResource: name, withExtension: "mp3") ??
                        Bundle.main.url(forResource: name, withExtension: "m4a") ??
                        Bundle.main.url(forResource: "success_generic", withExtension: "mp3") else {
            // Audio file not found - silent fail
            return
        }

        do {
            audioPlayer = try AVAudioPlayer(contentsOf: url)
            audioPlayer?.prepareToPlay()
            audioPlayer?.play()
        } catch {
            print("Failed to play audio: \(error)")
        }
    }

    private func getSuccessAudioName(for type: ObservationType) -> String {
        switch type {
        case .bloodPressure:
            return "success_blood_pressure"
        case .bloodGlucose:
            return "success_glucose"
        case .bodyTemperature:
            return "success_temperature"
        case .bodyWeight:
            return "success_weight"
        case .heartRate:
            return "success_pulse"
        case .oxygenSaturation:
            return "success_spo2"
        default:
            return "success_generic"
        }
    }

    // MARK: - Cleanup

    func stop() {
        audioPlayer?.stop()
        audioPlayer = nil
    }
}

File: android/app/src/main/java/com/carelog/voice/VoiceAcknowledgementPlayer.kt
package com.carelog.voice

import android.content.Context
import android.media.AudioAttributes
import android.media.MediaPlayer
import com.carelog.fhir.client.ObservationType
import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Plays voice acknowledgement audio clips for vital logging events.
 *
 * Audio plays through media stream to work even in silent mode.
 * Clips should be bundled in res/raw directory.
 */
@Singleton
class VoiceAcknowledgementPlayer @Inject constructor(
    @ApplicationContext private val context: Context
) {
    private var mediaPlayer: MediaPlayer? = null

    /**
     * Play success acknowledgement for a specific vital type.
     */
    fun playSuccess(type: ObservationType) {
        val resourceId = getSuccessResourceId(type)
        playAudio(resourceId)
    }

    /**
     * Play generic success acknowledgement.
     */
    fun playGenericSuccess() {
        // R.raw.success_generic - "Saved successfully"
        playAudio(getGenericSuccessResourceId())
    }

    /**
     * Play failure acknowledgement.
     */
    fun playFailure() {
        // R.raw.failure - "Could not save, please try again"
        playAudio(getFailureResourceId())
    }

    /**
     * Play no network acknowledgement.
     */
    fun playNoNetwork() {
        // R.raw.no_network - "No internet connection"
        playAudio(getNoNetworkResourceId())
    }

    /**
     * Play upload success acknowledgement.
     */
    fun playUploadSuccess() {
        // R.raw.upload_success - "File uploaded successfully"
        playAudio(getUploadSuccessResourceId())
    }

    private fun playAudio(resourceId: Int) {
        // Release any existing player
        mediaPlayer?.release()

        try {
            mediaPlayer = MediaPlayer().apply {
                setAudioAttributes(
                    AudioAttributes.Builder()
                        .setUsage(AudioAttributes.USAGE_MEDIA)
                        .setContentType(AudioAttributes.CONTENT_TYPE_SPEECH)
                        .build()
                )

                // Use AssetFileDescriptor for raw resources
                val afd = context.resources.openRawResourceFd(resourceId)
                if (afd != null) {
                    setDataSource(afd.fileDescriptor, afd.startOffset, afd.length)
                    afd.close()
                    prepare()
                    start()

                    setOnCompletionListener { mp ->
                        mp.release()
                        mediaPlayer = null
                    }
                }
            }
        } catch (e: Exception) {
            // Audio playback failed - log but don't crash
            mediaPlayer?.release()
            mediaPlayer = null
        }
    }

    private fun getSuccessResourceId(type: ObservationType): Int {
        // Map observation type to audio resource
        // These would be actual R.raw.xxx resource IDs
        return when (type) {
            ObservationType.BLOOD_PRESSURE -> getResourceIdForName("success_blood_pressure")
            ObservationType.BLOOD_GLUCOSE -> getResourceIdForName("success_glucose")
            ObservationType.BODY_TEMPERATURE -> getResourceIdForName("success_temperature")
            ObservationType.BODY_WEIGHT -> getResourceIdForName("success_weight")
            ObservationType.HEART_RATE -> getResourceIdForName("success_pulse")
            ObservationType.OXYGEN_SATURATION -> getResourceIdForName("success_spo2")
            else -> getGenericSuccessResourceId()
        }
    }

    private fun getGenericSuccessResourceId(): Int {
        return getResourceIdForName("success_generic")
    }

    private fun getFailureResourceId(): Int {
        return getResourceIdForName("failure")
    }

    private fun getNoNetworkResourceId(): Int {
        return getResourceIdForName("no_network")
    }

    private fun getUploadSuccessResourceId(): Int {
        return getResourceIdForName("upload_success")
    }

    private fun getResourceIdForName(name: String): Int {
        // Get resource ID by name, falling back to a default
        val resourceId = context.resources.getIdentifier(name, "raw", context.packageName)
        return if (resourceId != 0) resourceId else getFallbackResourceId()
    }

    private fun getFallbackResourceId(): Int {
        // Return a fallback resource ID or 0 if none exists
        return context.resources.getIdentifier("success_generic", "raw", context.packageName)
    }

    fun release() {
        mediaPlayer?.release()
        mediaPlayer = null
    }
}
